File: 200-number-of-islands/number-of-islands.cpp
class Solution {
public:
    
    void dfs(int i, int j, int n, int m, vector<vector<char>>& grid, vector<vector<bool>>&visited)
    {
        if(visited[i][j] == true)
        {
            return;
        }
        visited[i][j] = true;

        if(grid[i][j] == '0')
        {
            return;
        }
        if(j+1 < m) dfs(i, j+1, n, m, grid,visited);
        if(j-1 >= 0) dfs(i, j-1, n, m, grid,visited);
        if(i-1 >= 0) dfs(i-1, j, n, m, grid,visited);
        if(i+1 < n) dfs(i+1, j, n, m, grid,visited);

        return;
    }

    int numIslands(vector<vector<char>>& grid) 
    {
        int n=grid.size();
        int m=grid[0].size();
        int count =0;
        vector<vector<bool>> visited(n, vector<bool>(m, false));

        for(int i=0; i<n; i++)
        {
            for(int j=0; j<m; j++)
            {
                if(grid[i][j] == '1' && visited[i][j] == false)
                {
                    dfs(i, j, n, m, grid,visited);
                    count++;
                }
            }
        }
        return count;
    }
};

File: 200-number-of-islands/number-of-islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        m = len(grid)-1
        n = len(grid[0])-1
        seen = [[False for _ in range(n+1)] for _ in range(m+1)]
        ans = 0

        def island(i, j, m, n):
            if seen[i][j] == True:
                return
            else:
                seen[i][j] = True

            if grid[i][j] == '0':
                return

            if i+1<=m:
                island(i+1, j, m, n)
            if i-1>=0:
                island(i-1, j, m, n)
            if j+1<=n:
                island(i, j+1, m, n)
            if j-1>=0:
                island(i, j-1, m, n)
            return 1

        i, j = 0,0
        for i in range(m+1):
            for j in range(n+1):
                if seen[i][j] == False and grid[i][j] == '1':
                    ans+=1
                    island(i, j, m, n)

        return ans

